app/src/main/java/com/alphawallet/app/service/AssetDefinitionService.java

```diff
@@ -1203,6 +1238,11 @@ public StringBuilder getTokenAttrs(Token token, BigInteger tokenId, int count)
         TokenScriptResult.addPair(attrs, "tokenId", tokenId);
         TokenScriptResult.addPair(attrs, "ownerAddress", token.getWallet());
 
```

Comment:
This should be done with OO abstraction, also don't cast to ERC721 token just pass the token directly.

You could do something like this at the base level:

```
if (token instanceof ERC721Token) 
{
   addOpenSeaAttributes(attrs, token, tokenId);
}
```
Add the function getAssetForToken into class Token, return null, then specify it's an @Override in the ERC721Token definition.